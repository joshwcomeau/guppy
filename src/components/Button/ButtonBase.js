// @flow
import React, { Component } from 'react';
import styled from 'styled-components';

import { RAW_COLORS, COLORS } from '../../constants';

type Size = 'xsmall' | 'small' | 'medium' | 'large';

type Props = {
  size: Size,
  background?: string,
  hoverBackground?: string,
  textColor?: string,
  noPadding?: boolean,
  activeSplat?: boolean,
  disabled?: boolean,
  children: React$Node,
};

class ButtonBase extends Component<Props> {
  static defaultProps = {
    size: 'medium',
    background: RAW_COLORS.gray[200],
    textColor: COLORS.text,
  };

  getButtonElem = (size: Size) => {
    switch (size) {
      case 'xsmall':
        return XSmallButton;
      case 'small':
        return SmallButton;
      default:
      case 'medium':
        return MediumButton;
      case 'large':
        return LargeButton;
    }
  };

  render() {
    const {
      size,
      children,
      background,
      hoverBackground,
      textColor,
      disabled,
      ...delegated
    } = this.props;

    const Elem = this.getButtonElem(size);

    return (
      <Elem
        background={background}
        hoverBackground={hoverBackground}
        textColor={textColor}
        disabled={disabled}
        {...delegated}
      >
        {children}
      </Elem>
    );
  }
}

const ButtonBaseStyles = styled.button`
  position: relative;
  border: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  background: ${props => props.background};
  color: ${props => props.textColor};
  cursor: pointer;
  outline: none;
  white-space: nowrap;

  &:hover {
    background: ${props => props.hoverBackground};
  }

  &:disabled {
    filter: grayscale(100%);
    opacity: 0.75;
    cursor: initial;
  }

  &:not(:disabled):active {
    transform-origin: center center;
    transform: ${props => props.activeSplat && 'scale(1.1)'};
  }
`;

export const XSmallButton = styled(ButtonBaseStyles)`
  padding: ${props => (props.noPadding ? '0px' : '0px 12px')};
  height: ${props => (props.noPadding ? 'auto' : '22px')};
  border-radius: 15px;
  font-size: 12px;
`;

export const SmallButton = styled(ButtonBaseStyles)`
  padding: ${props => (props.noPadding ? '0px' : '0px 14px')};
  height: ${props => (props.noPadding ? 'auto' : '30px')};
  border-radius: 17px;
  font-size: 14px;
`;

export const MediumButton = styled(ButtonBaseStyles)`
  padding: ${props => (props.noPadding ? '0px' : '0px 20px')};
  height: ${props => (props.noPadding ? 'auto' : '38px')};
  border-radius: 19px;
  font-size: 16px;
`;

export const LargeButton = styled(ButtonBaseStyles)`
  padding: ${props => (props.noPadding ? '0px' : '0 32px')};
  height: ${props => (props.noPadding ? 'auto' : '48px')};
  border-radius: 24px;
  font-size: 24px;
`;

export default ButtonBase;
