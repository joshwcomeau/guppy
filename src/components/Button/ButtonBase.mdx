---
name: ButtonBase
menu: Reusable Components
---

import { Playground, Props } from 'docz';
import ButtonBase from './ButtonBase';

# ButtonBase

> Note: You probably don't want to use this component directly. It's just the
> underlying shared bit of code between FillButton and StrokeButton.

<Props of={ButtonBase} />

## Basic usage

<Playground>
  <ButtonBase>Default</ButtonBase>
</Playground>
