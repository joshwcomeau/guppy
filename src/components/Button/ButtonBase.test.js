/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import { shallow, mount } from 'enzyme';
import ButtonBase, {
  XSmallButton as xsmall,
  SmallButton as small,
  MediumButton as medium,
  LargeButton as large,
} from './ButtonBase';

const sizeComponents = {
  xsmall,
  small,
  medium,
  large,
};

describe('ButtonBase component', () => {
  describe('Button with size prop', () => {
    const wrapper = shallow(<ButtonBase />);
    const instance = wrapper.instance();

    for (const size in sizeComponents) {
      it(`should return sized button component for size = ${size}`, () => {
        const component = instance.getButtonElem(size);
        expect(component).toBe(sizeComponents[size]);
      });
      it(`should render button for ${size} size`, () => {
        // Using mount so the styles are available in the snapshot
        const wrapperSizedButton = mount(<ButtonBase size={size} />);
        expect(wrapperSizedButton).toMatchSnapshot();
      });
      it(`should render button for ${size} size with-out padding`, () => {
        // Using mount so the styles are available in the snapshot
        const wrapperSizedButton = mount(<ButtonBase size={size} noPadding />);
        expect(wrapperSizedButton).toMatchSnapshot();
      });
    }

    it('should return default size medium', () => {
      const componentDefault = instance.getButtonElem();
      expect(componentDefault).toBe(medium);
    });
  });
});
