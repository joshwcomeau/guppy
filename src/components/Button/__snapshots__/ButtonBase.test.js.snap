// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonBase component Button with size prop should render button for large size 1`] = `
.c1 {
  position: relative;
  border: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #eaeaea;
  color: #111111;
  cursor: pointer;
  outline: none;
  white-space: nowrap;
}

.c1:disabled {
  -webkit-filter: grayscale(100%);
  filter: grayscale(100%);
  opacity: 0.75;
  cursor: initial;
}

.c1:not(:disabled):active {
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

.c0 {
  padding: 0 32px;
  height: 48px;
  border-radius: 24px;
  font-size: 24px;
}

<ButtonBase
  background="#eaeaea"
  size="large"
  textColor="#111111"
>
  <Styled(styled.button)
    background="#eaeaea"
    textColor="#111111"
  >
    <styled.button
      background="#eaeaea"
      className="c0"
      textColor="#111111"
    >
      <button
        className="c0 c1"
      />
    </styled.button>
  </Styled(styled.button)>
</ButtonBase>
`;

exports[`ButtonBase component Button with size prop should render button for large size with-out padding 1`] = `
.c1 {
  position: relative;
  border: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #eaeaea;
  color: #111111;
  cursor: pointer;
  outline: none;
  white-space: nowrap;
}

.c1:disabled {
  -webkit-filter: grayscale(100%);
  filter: grayscale(100%);
  opacity: 0.75;
  cursor: initial;
}

.c1:not(:disabled):active {
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

.c0 {
  padding: 0px;
  height: auto;
  border-radius: 24px;
  font-size: 24px;
}

<ButtonBase
  background="#eaeaea"
  noPadding={true}
  size="large"
  textColor="#111111"
>
  <Styled(styled.button)
    background="#eaeaea"
    noPadding={true}
    textColor="#111111"
  >
    <styled.button
      background="#eaeaea"
      className="c0"
      noPadding={true}
      textColor="#111111"
    >
      <button
        className="c0 c1"
      />
    </styled.button>
  </Styled(styled.button)>
</ButtonBase>
`;

exports[`ButtonBase component Button with size prop should render button for medium size 1`] = `
.c1 {
  position: relative;
  border: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #eaeaea;
  color: #111111;
  cursor: pointer;
  outline: none;
  white-space: nowrap;
}

.c1:disabled {
  -webkit-filter: grayscale(100%);
  filter: grayscale(100%);
  opacity: 0.75;
  cursor: initial;
}

.c1:not(:disabled):active {
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

.c0 {
  padding: 0px 20px;
  height: 38px;
  border-radius: 19px;
  font-size: 16px;
}

<ButtonBase
  background="#eaeaea"
  size="medium"
  textColor="#111111"
>
  <Styled(styled.button)
    background="#eaeaea"
    textColor="#111111"
  >
    <styled.button
      background="#eaeaea"
      className="c0"
      textColor="#111111"
    >
      <button
        className="c0 c1"
      />
    </styled.button>
  </Styled(styled.button)>
</ButtonBase>
`;

exports[`ButtonBase component Button with size prop should render button for medium size with-out padding 1`] = `
.c1 {
  position: relative;
  border: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #eaeaea;
  color: #111111;
  cursor: pointer;
  outline: none;
  white-space: nowrap;
}

.c1:disabled {
  -webkit-filter: grayscale(100%);
  filter: grayscale(100%);
  opacity: 0.75;
  cursor: initial;
}

.c1:not(:disabled):active {
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

.c0 {
  padding: 0px;
  height: auto;
  border-radius: 19px;
  font-size: 16px;
}

<ButtonBase
  background="#eaeaea"
  noPadding={true}
  size="medium"
  textColor="#111111"
>
  <Styled(styled.button)
    background="#eaeaea"
    noPadding={true}
    textColor="#111111"
  >
    <styled.button
      background="#eaeaea"
      className="c0"
      noPadding={true}
      textColor="#111111"
    >
      <button
        className="c0 c1"
      />
    </styled.button>
  </Styled(styled.button)>
</ButtonBase>
`;

exports[`ButtonBase component Button with size prop should render button for small size 1`] = `
.c1 {
  position: relative;
  border: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #eaeaea;
  color: #111111;
  cursor: pointer;
  outline: none;
  white-space: nowrap;
}

.c1:disabled {
  -webkit-filter: grayscale(100%);
  filter: grayscale(100%);
  opacity: 0.75;
  cursor: initial;
}

.c1:not(:disabled):active {
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

.c0 {
  padding: 0px 14px;
  height: 30px;
  border-radius: 17px;
  font-size: 14px;
}

<ButtonBase
  background="#eaeaea"
  size="small"
  textColor="#111111"
>
  <Styled(styled.button)
    background="#eaeaea"
    textColor="#111111"
  >
    <styled.button
      background="#eaeaea"
      className="c0"
      textColor="#111111"
    >
      <button
        className="c0 c1"
      />
    </styled.button>
  </Styled(styled.button)>
</ButtonBase>
`;

exports[`ButtonBase component Button with size prop should render button for small size with-out padding 1`] = `
.c1 {
  position: relative;
  border: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #eaeaea;
  color: #111111;
  cursor: pointer;
  outline: none;
  white-space: nowrap;
}

.c1:disabled {
  -webkit-filter: grayscale(100%);
  filter: grayscale(100%);
  opacity: 0.75;
  cursor: initial;
}

.c1:not(:disabled):active {
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

.c0 {
  padding: 0px;
  height: auto;
  border-radius: 17px;
  font-size: 14px;
}

<ButtonBase
  background="#eaeaea"
  noPadding={true}
  size="small"
  textColor="#111111"
>
  <Styled(styled.button)
    background="#eaeaea"
    noPadding={true}
    textColor="#111111"
  >
    <styled.button
      background="#eaeaea"
      className="c0"
      noPadding={true}
      textColor="#111111"
    >
      <button
        className="c0 c1"
      />
    </styled.button>
  </Styled(styled.button)>
</ButtonBase>
`;

exports[`ButtonBase component Button with size prop should render button for xsmall size 1`] = `
.c1 {
  position: relative;
  border: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #eaeaea;
  color: #111111;
  cursor: pointer;
  outline: none;
  white-space: nowrap;
}

.c1:disabled {
  -webkit-filter: grayscale(100%);
  filter: grayscale(100%);
  opacity: 0.75;
  cursor: initial;
}

.c1:not(:disabled):active {
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

.c0 {
  padding: 0px 12px;
  height: 22px;
  border-radius: 15px;
  font-size: 12px;
}

<ButtonBase
  background="#eaeaea"
  size="xsmall"
  textColor="#111111"
>
  <Styled(styled.button)
    background="#eaeaea"
    textColor="#111111"
  >
    <styled.button
      background="#eaeaea"
      className="c0"
      textColor="#111111"
    >
      <button
        className="c0 c1"
      />
    </styled.button>
  </Styled(styled.button)>
</ButtonBase>
`;

exports[`ButtonBase component Button with size prop should render button for xsmall size with-out padding 1`] = `
.c1 {
  position: relative;
  border: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #eaeaea;
  color: #111111;
  cursor: pointer;
  outline: none;
  white-space: nowrap;
}

.c1:disabled {
  -webkit-filter: grayscale(100%);
  filter: grayscale(100%);
  opacity: 0.75;
  cursor: initial;
}

.c1:not(:disabled):active {
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

.c0 {
  padding: 0px;
  height: auto;
  border-radius: 15px;
  font-size: 12px;
}

<ButtonBase
  background="#eaeaea"
  noPadding={true}
  size="xsmall"
  textColor="#111111"
>
  <Styled(styled.button)
    background="#eaeaea"
    noPadding={true}
    textColor="#111111"
  >
    <styled.button
      background="#eaeaea"
      className="c0"
      noPadding={true}
      textColor="#111111"
    >
      <button
        className="c0 c1"
      />
    </styled.button>
  </Styled(styled.button)>
</ButtonBase>
`;
